### PR TITLE
NuGet.Client builds: Don't use Visual Studio version in bin, obj folder paths 

### PR DESCRIPTION
**Pull request by zivkan**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/1014

Regression? No

## Description

For years, when you build NuGet.Client, it used the VS version in the bin/obj folder path. This PR will:

* Change build.ps1 to tell MSBuild to use all CPU cores when building `build\build.proj`, rather than just 1.  Although, I'm not sure this actually worked for some reason. I feel like building NuGet.sln is much faster than build\build.proj.
* Remove VisualStudioVersion from obj/bin folder path
* Stop adding `-RTM` to folder path when BuildRTM is true
* There were a few tests that looked for assemblies in bin folder. These no longer expect the Visual Studio version folder.
* Stop setting DOTNET_HOME and DOTNET_INSTALL_DIRECTORY in configure.ps1.
* Scripts look for msbuild on the PATH, and fall back to using vswhere.
* Change our InstallVSIX.ps1 script, used for E2E tests and Apex, to use vswhere to find the VS install path, to locate vsixinstaller.exe, no matter what the install directory is, and also find and pass through the VS instance ID, so theoretically it might install/downgrade one of the VS installs, not multiple. Although this didn't appear to actually improve at all.

There's probably more, but basically it's about making builds more reliable.

## PR Checklist

- [X] PR has a meaningful title
- [X] PR has a linked issue.
- [X] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [X] N/A: Build scripts

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [X] N/A

### CODE REVIEW

File: build.ps1 [hunk 1]
```diff
@@ -81,11 +81,6 @@ Trace-Log "Build #$BuildNumber started at $startTime"
 
 Test-BuildEnvironment -CI:$CI
 
-if (-not $VSToolsetInstalled) {
```
Comment (nkolev92) [inline build.ps1:84]: why not check for a toolset anymore?

File: build/common.project.props [hunk 2]
```diff
@@ -113,10 +112,6 @@
     <VsixPublishDestination>$(ArtifactRoot)$(VsixOutputDirName)\</VsixPublishDestination>
   </PropertyGroup>
 
-  <PropertyGroup Condition=" '$(BuildRTM)' == 'true' ">
```
Comment (nkolev92) [inline build/common.project.props:116]: in every case, rtm and non-trm builds will have the same output directory right?

File: build/common.ps1 [hunk 3]
```diff
@@ -122,6 +122,7 @@ Function Invoke-BuildStep {
                 Trace-Log "[STOPPED +$(Format-ElapsedTime $sw.Elapsed)] $BuildStep"
             }
             else {
+                $err
```
Comment (nkolev92) [inline build/common.ps1]: what this do? this doesn't seem like a known variable, I only found $Error

File: build.ps1 (hunk 1 above)
Comment (nkolev92) [reply build.ps1:84]: Figured out toolsets are no more. Yay :)

It won't affect my workflow, but worth notifying others of this change.

File: scripts/cibuild/ConfigureVstsBuild.ps1 [hunk 4]
```diff
@@ -65,10 +65,11 @@ Function Update-VsixVersion {
 
 Function Set-RtmLabel {
     param(
-        [string]$BuildRTM
+        [Parameter(Mandatory = $true)]
+        [boolean]$isRTMBuild
```
Comment (nkolev92) [inline scripts/cibuild/ConfigureVstsBuild.ps1:69]: Consider `Switch` instead.

File: scripts/e2etests/RunFunctionalTests.ps1 [hunk 5]
```diff
@@ -6,28 +6,23 @@ param (
...
     [Parameter(Mandatory=$true)]
-    [string]$NuGetDropPath,
-    [Parameter(Mandatory=$true)]
-    [string]$FuncTestRoot,
-    [Parameter(Mandatory=$true)]
-    [string]$RunCounter,
-    [Parameter(Mandatory=$true)]
-    [ValidateSet("17.0")]
-    [string]$VSVersion)
+    [string]$FuncTestRoot)
```
Comment (nkolev92) [inline scripts/e2etests/RunFunctionalTests.ps1]: We should probably port a version of these changes to release-5.11.x right?

File: scripts/e2etests/VSUtils.ps1 [hunk 6]
```diff
@@ -2,11 +2,44 @@ $VSInstallerProcessName = "VSIXInstaller"
 
 . "$PSScriptRoot\Utils.ps1"
 
+function Get-LatestVSInstance
```
Comment (nkolev92) [inline scripts/e2etests/VSUtils.ps1]: This name almost got me. Worth including the fact that it looks for the appropriate version baed on our config props?

SO, `Get-LatestCompatibleVSInstance` maybe?

**Review by nkolev92 — COMMENTED** [review]
These are some awesome changes!

Can you please make the approriate changes in https://github.com/NuGet/NuGet.Client/blob/dev/scripts/nuget-debug-helpers.ps1 as well!

File: build.ps1 (hunk 1 above)
Comment (zivkan) [reply build.ps1:84]: I never understood how the toolset/config file worked anyway. How might it affect other people's workflows? Auto-detecting msbuild on the PATH, and falling back to the latest that vswhere provides seems quite safe to me.

File: build/common.project.props (hunk 2 above)
Comment (zivkan) [reply build/common.project.props:116]: Yes, that's the goal. On CI we run the RTM and non-RTM builds on different jobs, so there's never a scenario where both builds need to be available side-by-side.  After talking to someone in dnceng, it appears that other teams manually queue a build of a specific commit, with a variable set to their equivalent of `BuildRTM`, so they don't even do dual RTM/nonRTM builds as we do.  I'm considering changing NuGet.Client in the same way, but not in this PR.

File: build/common.ps1 (hunk 3 above)
Comment (zivkan) [reply build/common.ps1]: I think it was an attempt at debugging before I figured out the catch block on lines 109-111, which was added in another PR which I actually started working on after I started this PR's branch.

File: scripts/e2etests/RunFunctionalTests.ps1 (hunk 5 above)
Comment (zivkan) [reply scripts/e2etests/RunFunctionalTests.ps1]: Yes, I think we'll need to, in order to get the servicing branch to run tests successfully while dev16 and dev17 are installed side-by-side on the CI agents.

File: scripts/e2etests/VSUtils.ps1 (hunk 6 above)
Comment (zivkan) [reply scripts/e2etests/VSUtils.ps1]: I agree, I started off writing it to get the latest version, but then changed it to as you described. I decided to go a different way, I kept the function the same, removed the config.props reading, and added a `VersionRange` parameter. I added a new `Get-VisualStudioVersionRangeFromConfig` function, and changed the 2 places that call `Get-LatestVSInstance` to call both methods. Maybe it's a little extra complicated, but they're useful utility functions, and make it a little easier to debug these scripts, I think.

File: build.ps1 (hunk 1 above)
Comment (nkolev92) [reply build.ps1:84]: Once you run configure.ps1, you'd be tied to that version of msbuild on *every* build.ps1 run regardless where it's coming from. 

So if your workflow is to *not* always use the latest msbuild, you *could* run into some problems. 

Worth running by the team.

File: scripts/cibuild/ConfigureVstsBuild.ps1 (hunk 4 above)
Comment (zivkan) [reply scripts/cibuild/ConfigureVstsBuild.ps1:69]: Looking at [the docs on switch parameters](https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/types-of-cmdlet-parameters?view=powershell-7.1#switch-parameters), I don't really like it.

As a boolean parameter, we can parse the string as a bool in the outer function, and then pass the strongly typed value into the inner function, which we can then use in a condition.

If we want to use a switch parameter, then the outer function needs to condition whether or not it passes the switch, as well as the inner function having a if-else block.

I think it just adds more branching.

**Comment by donnie-msft** [comment]
Anything useful we can add in https://github.com/NuGet/NuGet.Client/blob/dev/CONTRIBUTING.md to make it clear how to specify msbuild versions, and the implications of using one over another?

**Comment by zivkan** [comment]
Unless contributors are doing something "weird", I don't think it'll affect anyone. If they're already using "developer powershell" or "developer command prompt", it'll still work. If they're not, it should also still work.

In fact, with this PR, building NuGet.Client is more likely to work the same as contributor's own repositories than NuGet.Client's dev branch does currently, therefore decreasing the usefulness of any such documentation.

Therefore no, I don't think it's worth changing the contribution guide. `configure.ps1` and `build.ps1` are still the correct build steps.

**Review by dominoFire — APPROVED** [review]
Signing-off. If there is changes pending in other release pipelines due to changes on artifacts file paths, please create follow-up issues.

File: build/common.project.props (hunk 2 above)
Comment (nkolev92) [reply build/common.project.props:116]: Resolving this conversation.

File: scripts/cibuild/ConfigureVstsBuild.ps1 (hunk 4 above)
Comment (nkolev92) [reply scripts/cibuild/ConfigureVstsBuild.ps1:69]: I disagree, but I don't care too much :D

**Review by nkolev92 — APPROVED** [review]
LGTM.

**Pull request merged**